### PR TITLE
feat: NFTs timeline

### PR DESCRIPTION
**Pull request by LeifXu**
# Pull Request

## Related Issue

Closes FW-1206

## Checklist

-   [ ] I have tested these changes locally.
-   [ ] I have reviewed the code changes.
-   [ ] I have added unit tests if applicable.

### CODE REVIEW

File: src/components/Actions/NFTReportSpamButton.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    ref,
+) {
+    const [, reportSpamNFT] = useReportSpamNFT();
+    return (
+        <MenuButton
+            {...rest}
+            ref={ref}
+            onClick={async () => {
+                rest.onClick?.();
+                await reportSpamNFT(contractAddress);
```
Comment (guanbinrui) [inline src/components/Actions/NFTReportSpamButton.tsx:24]: Handling loading & error.

File: src/components/NFTDetail/ReportSpamButton.tsx [hunk 2]
```diff
@@ -1,38 +1,17 @@
...
-import { getSnackbarMessageFromError } from '@/helpers/getSnackbarMessageFromError.js';
-import { ConfirmModalRef } from '@/modals/controls.js';
-import { reportNFT } from '@/services/reportNFT.js';
+import { useReportSpamNFT } from '@/hooks/useReportSpamNFT.js';
 
 export function ReportSpamButton(props: { collectionId: string }) {
     const { collectionId } = props;
+    const [, reportSpamNFT] = useReportSpamNFT();
     return (
         <button
```
Comment (guanbinrui) [inline src/components/NFTDetail/ReportSpamButton.tsx:12]: `<ClickableButton />`

File: src/components/NFTs/NFTFeedAction.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                            ) : (
+                                <Trans>
+                                    <span className="font-bold uppercase text-main">Minted</span> an NFT for free
+                                </Trans>
+                            )}
+                        </div>
+                    </>
+                );
+            case NFTFeedTransAction.Transfer:
+                if (toAddress === ownerAddress) {
```
Comment (guanbinrui) [inline src/components/NFTs/NFTFeedAction.tsx:47]:
```suggestion
                if (isSameAddress(toAddress, ownerAddress)) {
```

File: src/components/Actions/NFTReportSpamButton.tsx (hunk 1 above)
Comment (LeifXu) [reply src/components/Actions/NFTReportSpamButton.tsx:24]: The loading and error have been handled in `useReportSpamNFT`

File: src/constants/enum.ts [hunk 4]
```diff
@@ -40,6 +41,7 @@ export enum SourceInURL {
     Twitter = 'twitter',
     Article = 'article',
     Wallet = 'wallet',
+    NFTs = 'nft',
```
Comment (guanbinrui) [inline src/constants/enum.ts:44]:
```suggestion
    NFTs = 'nfts',
```

File: src/store/useInvalidNFTStore.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    has: (contractAddress: ContractAddress, tokenId: TokenId, chainId: ChainId) => boolean;
+}
+
+export const useInvalidNFTStore = create<InvalidNFTStore, [['zustand/immer', never]]>(
+    immer((set, get) => ({
+        dataSet: {},
+        size: 0,
+        registerNotFoundKey: (contractAddress, tokenId, chainId) => {
+            return set((state) => {
+                state.dataSet[`${contractAddress}.${tokenId}.${chainId}`] = true;
```
Comment (guanbinrui) [inline src/store/useInvalidNFTStore.ts:22]: Recommend to use CAIP-10 as the key..

File: src/components/TokenPrice.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { formatBalance } from '@masknet/web3-shared-base';
```
Comment (guanbinrui) [inline src/components/TokenPrice.tsx:1]:
```suggestion
// eslint-disable-next-line no-restricted-imports
import { formatBalance } from '@/maskbook/packages/web3-shared/base/src/helpers/formatBalance.js';
```

File: src/components/NFTs/NFTFeedHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import { Avatar } from '@/components/Avatar.js';
+import { ChainIcon } from '@/components/NFTDetail/ChainIcon.js';
+import { NFTMoreAction } from '@/components/NFTs/NFTMoreAction.js';
+import { TimestampFormatter } from '@/components/TimeStampFormatter.js';
+import { SourceInURL } from '@/constants/enum.js';
+import { Link } from '@/esm/Link.js';
+import { classNames } from '@/helpers/classNames.js';
+import type { NFTOwnerDisplayInfo } from '@/providers/types/NFTs.js';
+
+interface NFTFeedHeaderProps {
```
Comment (guanbinrui) [inline src/components/NFTs/NFTFeedHeader.tsx:13]:
```suggestion
interface NFTFeedHeaderProps extends React.HTMLAttributes<HTMLDivElement> {
```

File: src/components/NFTs/NFTFeedBody.tsx [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                </div>
+                {/* used for layout calculation height */}
+                <div className="h-full w-full space-y-1.5 opacity-0 sm:hidden">
+                    <div className="aspect-square w-full" />
+                    <div className="h-[96px]" />
+                </div>
+            </div>
+            {tokenList.length > 1 ? (
+                <div className="mt-1.5 flex h-[18px] justify-between overflow-hidden text-lightSecond">
+                    <button
```
Comment (guanbinrui) [inline src/components/NFTs/NFTFeedBody.tsx:146]:
```suggestion
                    <ClickableButton
```

File: src/components/NFTs/DiscoverNFTList.tsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        },
+        initialPageParam: '',
+        getNextPageParam: (lastPage) => lastPage.nextIndicator?.id,
+        select: (data) => data.pages.flatMap((p) => p.data),
+    });
+
+    const invalidNFTCount = useInvalidNFTStore((state) => state.size);
+    const filteredData = useMemo(() => {
+        const invalidNFTStore = useInvalidNFTStore.getState();
+        return nftQueryResult.data.filter((feed) => {
```
Comment (guanbinrui) [inline src/components/NFTs/DiscoverNFTList.tsx:34]: Built into the select.

File: src/components/Actions/NFTReportSpamButton.tsx (hunk 1 above)
Comment (guanbinrui) [reply src/components/Actions/NFTReportSpamButton.tsx:24]: It only handled the error (presents error message). For loading, recommend to display a `Reportng...` message on the button.

File: src/hooks/useReportSpamNFT.tsx [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+            variant: 'normal',
+            content: (
+                <div className="text-main">
+                    <Trans>Are you sure you want to report this collection?</Trans>
+                </div>
+            ),
+        });
+        if (!confirmed) return;
+        try {
+            await reportNFT(collectionId);
```
Comment (guanbinrui) [inline src/hooks/useReportSpamNFT.tsx:22]: Add a success message.

**Pull request merged**